Expose dataStoreIdentifier and webPushPartition getters on WKWebsiteDataStore

bnham opened the pull request:
#### f8d4e263f3a9d2e00d6f348a7d176d57bcfd02f5
<pre>
Expose dataStoreIdentifier and webPushPartition getters on WKWebsiteDataStore
<a href="https://bugs.webkit.org/show_bug.cgi?id=252067">https://bugs.webkit.org/show_bug.cgi?id=252067</a>
rdar://problem/105287995

Reviewed by Sihui Liu.

When showing a notification, we need to provide the embedder with a way to know the data store
identifier and push partition that the notification is associated with. This patch facilitates this
by adding those properties to WKWebsiteDataStore. You could do this before, but it would require
accessing the configuration property on WKWebsiteDataStore, which does a large copy operation.

Additionally, this provides a way to go from a the C API WKWebsiteDataStoreRef to Obj-C API
WKWebsiteDataStore. This is because the callback for showing a notification uses the C API.

Contains miscellaneous unified build fixes.

* Source/WebKit/PlatformMac.cmake:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/C/mac/WKWebsiteDataStoreRefPrivateMac.h: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKInspectorInternal.h.
* Source/WebKit/UIProcess/API/C/mac/WKWebsiteDataStoreRefPrivateMac.mm: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKInspectorInternal.h.
(WKWebsiteDataStoreGetDataStore):
* Source/WebKit/UIProcess/API/Cocoa/APIAttachmentCocoa.mm:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _identifier]):
(-[WKWebsiteDataStore _webPushPartition]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKInspector.mm:
* Source/WebKit/UIProcess/API/Cocoa/_WKInspectorInternal.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
* Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::createWebsiteDataStoreAndPrepare):
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/260379@main">https://commits.webkit.org/260379@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/31f4b61bc920b891546c7165c3d9f3bd0d744fbf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108099 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17169 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40967 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117224 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116548 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111988 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18524 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8465 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100303 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113867 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/13970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97192 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41899 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95877 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28830 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10051 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30178 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10773 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7081 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16199 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49769 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7184 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12354 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1019,5 +1019,23 @@ -(void)_setBackupExclusionPeriodForTesting:(double)seconds completionHandler:(vo
+        return nil;
+
+    auto span = identifier->toSpan();
+    return adoptNS([[NSUUID alloc] initWithUUIDBytes:span.data()]).autorelease();
```
I think this is wrong because the stored `UUID` is converted with `UUID(NSUUID * nsUUID)`, i.e. from uuid string. It should be easily verified with existing test as in [260026@main](https://commits.webkit.org/260026@main).

bnham replied:
Ah I see didn't realize that UUID defines an NSUUID conversion. I'll use that instead.

bnham replied:
Sounds like there is also an issue with the way the endianness of the raw bytes is represented when passing toSpan into the NSUUID constructor. In any case now I've switched to using the operator.

szewai reviewed (approved):
I think we'd like to avoid adding new C API nowadays, but if it's necessary to avoid big changes then it's probably fine.

I would suggest adding a one-line change to the test in [260026@main](https://commits.webkit.org/260026@main) to ensure the uuid retrieved via the new SPI is the same as set, just in case our UUID implementation changes and causes regression.

bnham committed:
Expose dataStoreIdentifier and webPushPartition getters on WKWebsiteDataStore
https://bugs.webkit.org/show_bug.cgi?id=252067
rdar://problem/105287995

Reviewed by Sihui Liu.

When showing a notification, we need to provide the embedder with a way to know the data store
identifier and push partition that the notification is associated with. This patch facilitates this
by adding those properties to WKWebsiteDataStore. You could do this before, but it would require
accessing the configuration property on WKWebsiteDataStore, which does a large copy operation.

Additionally, this provides a way to go from a the C API WKWebsiteDataStoreRef to Obj-C API
WKWebsiteDataStore. This is because the callback for showing a notification uses the C API.

Contains miscellaneous unified build fixes.

* Source/WebKit/PlatformMac.cmake:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/C/mac/WKWebsiteDataStoreRefPrivateMac.h: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKInspectorInternal.h.
* Source/WebKit/UIProcess/API/C/mac/WKWebsiteDataStoreRefPrivateMac.mm: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKInspectorInternal.h.
(WKWebsiteDataStoreGetDataStore):
* Source/WebKit/UIProcess/API/Cocoa/APIAttachmentCocoa.mm:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _identifier]):
(-[WKWebsiteDataStore _webPushPartition]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKInspector.mm:
* Source/WebKit/UIProcess/API/Cocoa/_WKInspectorInternal.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
* Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::createWebsiteDataStoreAndPrepare):
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/260379@main
